integration/spark/src/main/common/java/io/openlineage/spark/api/AbstractQueryPlanDatasetBuilder.java

```diff
@@ -0,0 +1,129 @@
+      UnknownEntryFacetListener.getInstance();
+
+  private final boolean searchDependencies;
+
+  public AbstractQueryPlanDatasetBuilder(OpenLineageContext context, boolean searchDependencies) {
+    this.context = context;
+    this.searchDependencies = searchDependencies;
+  }
+
+  protected DatasetFactory<OpenLineage.OutputDataset> outputDataset() {
+    return DatasetFactory.output(context.getOpenLineage());
+  }
+
+  protected DatasetFactory<OpenLineage.InputDataset> inputDataset() {
+    return DatasetFactory.input(context.getOpenLineage());
+  }
+
+  public abstract List<D> apply(P p);
+
+  public List<D> apply(T event) {
+    return context
+        .getQueryExecution()
+        .map(
+            qe -> {
+              QueryPlanVisitor<LogicalPlan, D> visitor = asQueryPlanVisitor(event);
+              if (searchDependencies) {
+                return ScalaConversionUtils.fromSeq(qe.optimizedPlan().collect(visitor)).stream()
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.toList());
+              } else if (visitor.isDefinedAt(qe.optimizedPlan())) {
+                return visitor.apply(qe.optimizedPlan());
+              } else {
+                return Collections.<D>emptyList();
+              }
+            })
+        .orElseGet(Collections::emptyList);
+  }
+
+  public <L extends LogicalPlan> QueryPlanVisitor<L, D> asQueryPlanVisitor(T event) {
+    AbstractQueryPlanDatasetBuilder<T, P, D> builder = this;
+    return new QueryPlanVisitor<L, D>(context) {
+      @Override
+      public boolean isDefinedAt(LogicalPlan x) {
+        return builder.isDefinedAt(event) && isDefinedAtLogicalPlan(x);
+      }
+
+      @Override
+      public List<D> apply(LogicalPlan x) {
+        unknownEntryFacetListener.accept(x);
+        return builder.apply((P) x);
+      }
+    };
+  }
+
+  protected PartialFunction<LogicalPlan, Collection<D>> delegate(
+      Collection<PartialFunction<LogicalPlan, List<D>>> visitors,
+      Collection<? extends PartialFunction<Object, Collection<D>>> builders,
+      T event) {
+
+    return PlanUtils.merge(
+        Stream.concat(
+                visitors.stream(),
+                builders.stream()
+                    .filter(
+                        builder ->
+                            builder instanceof AbstractQueryPlanDatasetBuilder
+                                && builder.isDefinedAt(event))
+                    .map(
+                        builder ->
+                            ((AbstractQueryPlanDatasetBuilder<Object, LogicalPlan, D>) builder)
+                                .asQueryPlanVisitor(event)))
+            .collect(Collectors.toList()));
+  }
+
+  /**
+   * Similar to the logic in {@link AbstractPartial}, this reads the type from the <i>second</i>
+   * generic argument on the class, if it is present and non-null.
+   *
+   * @param logicalPlan
+   * @return
+   */
+  protected boolean isDefinedAtLogicalPlan(LogicalPlan logicalPlan) {
```

Comment:
I experienced some strange behavior of this method although not sure what was that. I think the behavior of this method may vary for extended classes with different generic types on the top level class. While extending class, one can omit some parametrized types or even change their ordering.

Consider making it abstract and pass implementation to extending classes, which is more explicit and is already done that way in our examples.

Reply:
I updated the tests to be explicit about this - the generic type arguments are on the _superclass_, not on the concrete implementation. So in your example, `extends AbstractQueryPlanDatasetBuilder<SparkListenerEvent, LogicalPlan, OutputDataset>` provides the concrete type argument. In this code, we look at the second argument, which is `LogicalPlan` in the example.

I did notice that we don't handle the case when some arguments are supplied, but others aren't. I added some handling for `TypeVariable`s, which also implement `Type`, but aren't `Class`es, thus cause a `ClassCastException` without explicit handling.